nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+
+func executePreupgradeChecks(d *schema.ResourceData, m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := nsx.NewUpgradeClient(connector)
+	err := client.Executepreupgradechecks(nil, nil, nil, nil, nil, nil)
+	if err != nil {
+		return err
+	}
+	for _, componentType := range precheckComponentTypes {
+		err = waitForPrecheckComplete(m, componentType)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getPrecheckErrors(m interface{}, typeParam string) ([]nsxModel.UpgradeCheckFailure, error) {
+	connector := getPolicyConnector(m)
+	client := pre_upgrade_checks.NewFailuresClient(connector)
+	resultList, err := client.List(nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, nil, &typeParam, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	return resultList.Results, nil
+}
+
+func setFailedPrechecksInSchema(d *schema.ResourceData, precheckWarnings []nsxModel.UpgradeCheckFailure) error {
+	var failedPrechecksList []map[string]interface{}
+	for _, warning := range precheckWarnings {
+		if *warning.NeedsAck {
+			id := *warning.Id
+			message := *(warning.Message.Message)
+			elem := make(map[string]interface{})
+			elem["id"] = id
+			elem["message"] = message
+			failedPrechecksList = append(failedPrechecksList, elem)
+		}
+	}
+	return d.Set("failed_prechecks", failedPrechecksList)
+}
+
+func waitForBundleUpload(m interface{}, bundleID string) error {
+	connector := getPolicyConnector(m)
+	client := bundles.NewUploadStatusClient(connector)
+	pendingStates := []string{
+		nsxModel.UpgradeBundleUploadStatus_STATUS_UPLOADING,
+		nsxModel.UpgradeBundleUploadStatus_STATUS_VERIFYING,
+	}
+	targetStates := []string{
+		nsxModel.UpgradeBundleUploadStatus_STATUS_SUCCESS,
+		nsxModel.UpgradeBundleUploadStatus_STATUS_FAILED,
+	}
+	timeout := 1200
+	stateConf := &resource.StateChangeConf{
+		Pending: pendingStates,
+		Target:  targetStates,
+		Refresh: func() (interface{}, string, error) {
+			state, err := client.Get(bundleID)
+			if err != nil {
+				msg := fmt.Sprintf("Error while retrieving upload status of bundle %s", bundleID)
+				return state, nsxModel.UpgradeBundleUploadStatus_STATUS_FAILED, logAPIError(msg, err)
+			}
+
+			log.Printf("[DEBUG] Current status for uploading bundle %s is %s", bundleID, *state.Status)
+
+			return state, *state.Status, nil
+		},
+		Timeout:    time.Duration(timeout) * time.Second,
+		MinTimeout: 1 * time.Second,
+		Delay:      1 * time.Second,
+	}
+	_, err := stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Failed to upload bundle %s: %s", bundleID, err)
+	}
+	return nil
+}
+
+func waitForUcUpgrade(m interface{}) error {
+	connector := getPolicyConnector(m)
+	client := upgrade.NewUcUpgradeStatusClient(connector)
+	pendingStates := []string{
+		nsxModel.UcUpgradeStatus_STATE_NOT_STARTED,
+		nsxModel.UcUpgradeStatus_STATE_IN_PROGRESS,
+	}
+	targetStates := []string{
+		nsxModel.UcUpgradeStatus_STATE_SUCCESS,
+		nsxModel.UcUpgradeStatus_STATE_FAILED,
+	}
+	timeout := 1200
```

Comment:
Should we make the timeout configurable on the resource?

Reply:
Done. now the timeout for upgrading UC, execute precheck and upload bundle configurable to user